phper/src/functions.rs

```diff
@@ -347,35 +562,71 @@ impl Argument {
         let name = ensure_end_with_zero(name);
         Self {
             name,
+            type_hint: None,
             pass_by_ref: true,
             required: false,
+            nullable: false,
+            default_value: None,
         }
     }
+
+    /// Add a type-hint to the argument
+    pub fn with_type_hint(mut self, type_hint: ArgumentTypeHint) -> Self {
+        self.type_hint = Some(type_hint);
+        self
+    }
+
+    /// Allow type-hint to be nullable
+    pub fn allow_null(mut self) -> Self {
+        self.nullable = true;
+        self
+    }
+
+    /// Argument is required (also see by_*<ref|val>)
+    pub fn required(mut self) -> Self {
+        self.required = true;
+        self
+    }
+
+    /// Argument is optional (also see by_<ref|val>_optional)
+    pub fn optional(mut self) -> Self {
```

Comment:
I think since this is a Breaking Change, we could remove `by_ref_optional` and `by_val_optional` and add `by_ref()` and `by_val` setters. So something like `Argument::new('name').by_ref().optional()`, with pass_by_ref:false being the default ?

Reply:
Allright!

Reply:
Do you want to make this change in this PR?

Reply:
Done!